### PR TITLE
[Doc] add gnn explainability tutorial

### PR DESCRIPTION
**Pull request by shenhao-stu**
## Description
- [X] create 7_captum_explainability.py

### CODE REVIEW

**Comment by mufeili** [comment]
@jermainewang For tutorial purpose, do we need to add `captum` to the dependency of `dgl`?

**Comment by mufeili** [comment]
@BarclayII It will be great if you can also take a look.

File: tutorials/blitz/7_captum_explainability.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+"""
+Explaining Graph Neural Network with Captum
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:2]:
```suggestion
Explaining Graph Neural Networks with Captum
```

**Comment by mufeili** [comment]
@zhjwy9343 , @classicsong , you might also be interested.

File: tutorials/blitz/7_captum_explainability.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+===========================================
+
+By the end of this tutorial, you will be able to:
+
+- Understand the basic concepts of GNN explainability
+- Use Captum to explain the node classification task
+- Find the most important nodes and edges to the model
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:9]: I will suggest replacing L8-9 with

```
- Use Captum to explain the predicted node classes of a GNN
```

File: tutorials/blitz/7_captum_explainability.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+- Visualize the node-centered subgraphs
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:10]:
```suggestion
- Visualize prediction explanations by plotting node-centered subgraphs
```

File: tutorials/blitz/7_captum_explainability.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+######################################################################
+# Overviews of GNN Explainability
+# -------------------------------
+# In recent years, research on the interpretability of deep learning models has
+# made significant progress. Compared to cv and nlp domains, there is less
+# research and application of graph model interpretability, yet it is the key
+# to understanding deep graph neural networks. Generally, GNN explanation
+# research often starts with the following tasks: ** Which input edges are more
+# important? Which input nodes are more important? **
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:24]: Note that interpretability and explainability are close but not exactly the same. The former one might be associated with white box models like linear regression and decision tree. I will suggest changing this paragraph to

```
Like other deep learning models, GNNs are black box models. This might hinder the deployment of GNNs in risk-sensitive scenarios. Explaining GNNs has been an important research problem. Many GNN explainability methods approach this problem by identifying important nodes and edges in the input graphs.
```

File: tutorials/blitz/7_captum_explainability.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+# Node-centered subgraphs play a critical role in analyzing GNNs. The k-hop
+# subgraph of a node fully determines the information a k-layer GNN exploits
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:27]:
```suggestion
# subgraph of a node fully determines the information that most k-layer GNNs exploit
```

File: tutorials/blitz/7_captum_explainability.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+# to generate its final node representation. Many GNN explanation methods
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:28]:
```suggestion
# to compute the associated node representation. Many GNN explanation methods
```

File: tutorials/blitz/7_captum_explainability.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+# provide explanations by extracting a subgraph and assigning importance
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:29]:
```suggestion
# provide an explanation by extracting a subgraph and assigning importance
```

File: tutorials/blitz/7_captum_explainability.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+# weights to the nodes and edges of it. We will visualize node-centered
+# weighted subgraphs through DGL's built-in functions. This is beneficial
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:31]:
```suggestion
# weighted subgraphs with a built-in DGL function. This is beneficial
```

File: tutorials/blitz/7_captum_explainability.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+# for debugging and understanding GNNs and GNN explanation methods.
+
+# For this demonstration, we will use IntegratedGradients from `Captum <https://github.com/pytorch/captum>`__ to
+# explain the predictions of a graph convolutional network (GCN).
+# Specifically, we try to find the most important nodes and edges to the model in node classification.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:36]:
```suggestion
# Specifically, we try to find the most important nodes and edges to the model in classifying nodes in a graph.
```

File: tutorials/blitz/7_captum_explainability.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+######################################################################
+# Loading Cora Dataset
+# --------------------
+#
+# First, we load DGL’s built-in Cora dataset and retrieve its graph structure, node labels (classes) and the number of node classes.
+# 
+
+# Install and import required packages.
+import dgl
+import dgl.data
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:54]: No need for this line

File: tutorials/blitz/7_captum_explainability.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+dataset = dgl.data.CoraGraphDataset()
+g = dataset[0]
+
+######################################################################
+# Define the model
+# ----------------
+# Then, we will build a two-layer Graph Convolutional Network (GCN).
+# Each layer computes new node representations by aggregating neighbor information.
+# What's more, we use GraphConv which supports ``edge_weight`` as a parameter to calculate
+# the importance of the edge in the edge explainability task.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:65]:
```suggestion
# the edge importance in model explanation
```

File: tutorials/blitz/7_captum_explainability.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+# 
+
+from dgl.nn import GraphConv
+# Define a class for GCN
+class GCN(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
+        super(GCN, self).__init__()
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:72]:
```suggestion
        super().__init__()
```

File: tutorials/blitz/7_captum_explainability.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+# Then, we will build a two-layer Graph Convolutional Network (GCN).
+# Each layer computes new node representations by aggregating neighbor information.
+# What's more, we use GraphConv which supports ``edge_weight`` as a parameter to calculate
+# the importance of the edge in the edge explainability task.
+# 
+
+from dgl.nn import GraphConv
+# Define a class for GCN
+class GCN(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:71]:
```suggestion
    def __init__(self, in_feats, num_classes, h_feats=16):
```

You can then remove `16` in `GCN(g.ndata['feat'].shape[1], 16, dataset.num_classes)`.

File: tutorials/blitz/7_captum_explainability.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model = GCN(g.ndata['feat'].shape[1], 16, dataset.num_classes).to(device)
+g = g.to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)
+
+features = g.ndata['feat']
+labels = g.ndata['label']
+train_mask = g.ndata['train_mask']
+
+for epoch in range(1,201):
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:100]:
```suggestion
for epoch in range(200):
```

File: tutorials/blitz/7_captum_explainability.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    model.train()
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:101]: Move this to before `for epoch in ...` since you do not call `model.eval()` within the for loop.

File: tutorials/blitz/7_captum_explainability.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    optimizer.zero_grad()
+    # Forward
+    logits = model(features, g)
+    
+    # Compute prediction
+    pred = logits.argmax(1)
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:107]: L106-107 are not used.

File: tutorials/blitz/7_captum_explainability.py [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    
+    # Compute loss
+    # Note that you should only compute the losses of the nodes in the training set.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:110]: This line may be not necessary as we can assume readers to already have some background on this topic.

File: tutorials/blitz/7_captum_explainability.py [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+    loss = F.cross_entropy(logits[train_mask], labels[train_mask])
+    
+    # Backward
+    loss.backward()
+    optimizer.step()
+
+######################################################################
+# Explaining the predictions
+# --------------------------
+# First, we will complete the task of which input nodes are more important.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:120]:
```suggestion
# We will explain why the model might make the prediction of class C for all nodes.
# C is the ground truth label of node 10 for demonstration purpose.
# 
# First, we assign importance to the nodes by attributing the prediction of class C
# for all nodes to the input node features with IntegratedGradients.
```

File: tutorials/blitz/7_captum_explainability.py [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+# We attribute the model predictions to the input node features with IntegratedGradients.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:121]: You can then remove this line.

File: tutorials/blitz/7_captum_explainability.py [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+#
+
+# Select the node with index 10 for interpretability analysis
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:124]:
```suggestion
# Select the node with index 10 for demonstration
```

File: tutorials/blitz/7_captum_explainability.py [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+output_idx = 10
+target = int(g.ndata['label'][output_idx])
+print(target)
+
+# Since the ``IntergratedGradients`` method only allows one argument to be passed, we use ``partial`` function to pass the default value to the forward function.
+#
+
+# import captum
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:132]: Remove this line.

File: tutorials/blitz/7_captum_explainability.py [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+from captum.attr import IntegratedGradients
+from functools import partial
+
+# Node explainability
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:136]: Remove this line.

File: tutorials/blitz/7_captum_explainability.py [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+ig = IntegratedGradients(partial(model.forward, g=g))
+# Attribute the predictions for node class 0 to the input features
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:138]:
```suggestion
# Attribute the predictions of node class C to the input node features
```

File: tutorials/blitz/7_captum_explainability.py [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+
+# We compute the node importance weights from the input feature weights and normalize them.
+#
+
+# Scale attributions to [0, 1]:
+ig_attr_node = ig_attr_node.abs().sum(dim=1)
+ig_attr_node /= ig_attr_node.max()
+
+
+# We visualize node-centered weighted subgraphs through DGL's built-in functions.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:152]:
```suggestion
# We visualize node-centered weighted subgraphs with a DGL built-in function.
```

File: tutorials/blitz/7_captum_explainability.py [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+
+# Visualize
+from utility import visualize_subgraph
+import matplotlib.pyplot as plt
+
+num_hops = 2
+ax, nx_g = visualize_subgraph(g, output_idx, num_hops, node_alpha=ig_attr_node)
+plt.show()
+
+# <img src="README.assets/Figure_1.png" style="zoom:50%;" />
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:163]: This line can be removed.

File: tutorials/blitz/7_captum_explainability.py [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+# Then, we will complete the task of which input edges are more important.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:164]:
```suggestion
# Then, we will similarly assign importance weights to edges.
```

File: tutorials/blitz/7_captum_explainability.py [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+# To apply the IntergratedGradients method, we redefine the forward function
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:165]: Remove this line.

File: tutorials/blitz/7_captum_explainability.py [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+#
+
+def model_forward(edge_mask, g):
+    out = model(g.ndata['feat'],g,edge_weight=edge_mask)
+    return out
+
+# Edge explainability
+edge_mask = torch.ones(g.num_edges()).requires_grad_(True).to(device)
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:173]: Add a comment

> Initialize an edge mask so that we can attribute to it.

File: tutorials/blitz/7_captum_explainability.py [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+from dgl.nn import GraphConv
+# Define a class for GCN
+class GCN(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
+        super(GCN, self).__init__()
+        self.conv1 = GraphConv(in_feats, h_feats)
+        self.conv2 = GraphConv(h_feats, num_classes)
+
+    def forward(self, in_feat, g, edge_weight=None):
+        h = self.conv1(g, in_feat, edge_weight=edge_weight)
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:77]: Add a comment above

> edge_weight will weight the messages in message passing for edges

File: tutorials/blitz/7_captum_explainability.py (hunk 28 above)
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:173]: What will happen if `requires_grad_(True)` is not specified?

File: tutorials/blitz/7_captum_explainability.py [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+
+def model_forward(edge_mask, g):
+    out = model(g.ndata['feat'],g,edge_weight=edge_mask)
+    return out
+
+# Edge explainability
+edge_mask = torch.ones(g.num_edges()).requires_grad_(True).to(device)
+ig = IntegratedGradients(partial(model_forward, g=g))
+ig_attr_edge = ig.attribute(edge_mask, target=target,
+                            internal_batch_size=g.num_nodes(), n_steps=50)
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:176]: Should `internal_batch_size` still be `g.num_nodes()`? Do we need to pass a value for this argument?

File: tutorials/blitz/7_captum_explainability.py [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+print(ig_attr_edge.shape)
+
+# We compute the node importance weights from the input feature weights and normalize them.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:179]:
```suggestion
# We compute the edge importance weights from the input feature weights and normalize them.
```

Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:179]: Remove this line.

File: tutorials/blitz/7_captum_explainability.py [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+#
+
+# Scale attributions to [0, 1]:
+ig_attr_edge = ig_attr_edge.abs()
+ig_attr_edge /= ig_attr_edge.max()
+
+# We visualize node-centered weighted subgraphs through DGL's built-in functions.
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:186]:
```suggestion
# We visualize a node-centered subgraph with edge weights
```

File: tutorials/blitz/7_captum_explainability.py [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+#
+
+# Visualize
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:189]: Remove this line

File: tutorials/blitz/7_captum_explainability.py [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+ax, nx_g = visualize_subgraph(g, output_idx, num_hops, edge_alpha=ig_attr_edge)
+plt.show()
+
+# <img src="README.assets/Figure_2.png" style="zoom:50%;" />
+#
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:194]: Remove L193-194

File: tutorials/blitz/7_captum_explainability.py [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+
+# Visualize node and edge explainability
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:196]:
```suggestion
# Visualize both node and edge weights
```

File: tutorials/blitz/7_captum_explainability.py [hunk 36]
```diff
@@ -0,0 +1,203 @@
...
+ax, nx_g = visualize_subgraph(g, output_idx, num_hops, node_alpha=ig_attr_node, edge_alpha=ig_attr_edge)
+plt.show()
+
```
Comment (mufeili) [inline tutorials/blitz/7_captum_explainability.py:199]: Remove the lines below.
